### PR TITLE
feat: iOS support for testplans

### LINKED ISSUE #685: iOS: Support .xctestrun from Test Plan
**XCode Version**: 13.3.1
**Flank Version**: v8.1.0

**Description:**

it would be very useful to be able to use Test Plans with Flank. 
At the moment getting the following error after building the UITestScheme and specifying .xctestrun of the Test Plan:

```
java.lang.ClassCastException: com.dd.plist.NSArray cannot be cast to com.dd.plist.NSDictionary
	at ftl.ios.Xctestrun.findTestNames(Xctestrun.kt:87)
	at ftl.ios.Xctestrun.findTestNames(Xctestrun.kt:73)
	at ftl.args.IosArgs$testShardChunks$2.invoke(IosArgs.kt:71)
	at ftl.args.IosArgs$testShardChunks$2.invoke(IosArgs.kt:28)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at ftl.args.IosArgs.getTestShardChunks(IosArgs.kt)
	at ftl.run.IosTestRunner$runTests$2.invokeSuspend(IosTestRunner.kt:44)
	at ftl.run.IosTestRunner$runTests$2.invoke(IosTestRunner.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
	at kotlinx.coroutines.CoroutineScopeKt.coroutineScope(CoroutineScope.kt:180)
	at ftl.run.IosTestRunner.runTests(IosTestRunner.kt:25)
	at ftl.run.TestRunner.runTests(TestRunner.kt:56)
	at ftl.run.TestRunner.newRun(TestRunner.kt:343)
	at ftl.cli.firebase.test.ios.IosRunCommand$run$1.invokeSuspend(IosRunCommand.kt:43)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:270)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:79)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at ftl.cli.firebase.test.ios.IosRunCommand.run(IosRunCommand.kt:42)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1729)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2101)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2068)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1935)
	at picocli.CommandLine.execute(CommandLine.java:1864)
	at ftl.Main$Companion.main(Main.kt:41)
	at ftl.Main.main(Main.kt)
```

Thank you!

### PR DESCRIPTION
**Pull request by jan-gogo**
Fixes #685 

## Documentation
https://github.com/Flank/flank/blob/master/docs/feature/ios_test_plans.md

From repository root:
```
. .env
flankScripts testArtifacts -b master resolve
flankScripts shell buildFlank
cd test_runner
flank ios run -c=./src/test/kotlin/ftl/fixtures/test_app_cases/flank-xctestrunv2-all.yml
```
Feel free to modify `flank-xctestrunv2-all.yml` to test possible edge-cases.

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

**Comment by jan-gogo** [comment]
Example `ios_shards.json` generated from `AllTests_AllTests_iphoneos13.7-arm64e.xctestrun`
```
{
  "en": {
    "UITests": [
      [
        "UITests/test1PLLocale"
      ],
      [
        "UITests/test1ENLocale"
      ],
      [
        "UITests/test1",
        "UITests/test3"
      ]
    ],
    "SecondUITests": [
      [
        "SecondUITests/test2ENLocale"
      ],
      [
        "SecondUITests/test2PLLocale",
        "SecondUITests/test13"
      ],
      [
        "SecondUITests/test11",
        "SecondUITests/test12"
      ]
    ]
  },
  "pl": {
    "UITests": [
      [
        "UITests/test1PLLocale"
      ],
      [
        "UITests/test1ENLocale"
      ],
      [
        "UITests/test1",
        "UITests/test3"
      ]
    ],
    "SecondUITests": [
      [
        "SecondUITests/test2ENLocale"
      ],
      [
        "SecondUITests/test2PLLocale",
        "SecondUITests/test13"
      ],
      [
        "SecondUITests/test11",
        "SecondUITests/test12"
      ]
    ]
  }
}
```

File: gradle.properties [hunk 1]
```diff
@@ -13,4 +13,4 @@ org.gradle.jvmargs=-Xmx2560m -XX:MaxPermSize=256m -Dfile.encoding=UTF-8
 kotlin.code.style=official
 org.gradle.parallel=true
 org.gradle.daemon=true
-org.gradle.caching=true
+org.gradle.caching=false
```
Comment (piotradamczyk5) [inline gradle.properties:16]: why disabling cache?

File: test_runner/src/main/kotlin/ftl/ios/xctest/FindXcTestNamesV2.kt [hunk 2]
```diff
@@ -1,27 +1,26 @@
...
+            configDict
+                .getTestTargets()
+                .associateBy { targetDict -> targetDict.getBlueprintName() }
+                .mapValues { (name, dict) ->
+                    findTestsForTarget(
+                        testRoot = xcTestRoot,
+                        testTargetName = name,
+                        testTargetDict = dict,
+                    )
+                }
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/ios/xctest/FindXcTestNamesV2.kt:25]: I would move it to new function to avoid nested streams

File: test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV1.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+fun NSDictionary.reduceXcTestRunV1(
+    targets: Map<String, List<String>>,
+): NSDictionary = apply {
+    (keys - targets.keys - XCTEST_METADATA).forEach(this::remove)
+    targets.forEach { (testTarget, methods) ->
+        set(
+            testTarget,
+            getValue(testTarget)
+                .let { it as NSDictionary }
+                .setOnlyTestIdentifiers(methods)
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV1.kt:16]: same here I would move it to new function

File: test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+): NSDictionary = apply {
+    set(
+        TEST_TARGETS,
+        getTestTargets()
+            .mapNotNull {
+                targets[
+                    it.getBlueprintName()
+                ]?.let { methods ->
+                    it.setOnlyTestIdentifiers(methods)
+                }
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt:37]:
```kotlin
targets[ it.getBlueprintName()]
    ?.let { methods ->it.setOnlyTestIdentifiers(methods)}
```

File: test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    val shardTargets = shardChunks.testCases.map { shardMethods ->
+        targets.mapValues { (_, methods) ->
+            methods.intersect(shardMethods).toList()
+        }.filterValues {
+            it.isNotEmpty()
+        }
+    }
+
+    return shardChunks to shardTargets
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt:88]: we could move this 2 vals to functions and then return
``` 
private fun IosArgs.calculateConfigurationShards(
    targets: Map<String, List<String>>,
    regexList: List<Regex>
): Pair<List<Chunk>, List<XctestrunMethods>> = newFunction1() to newFunction2()
```

File: test_runner/src/test/kotlin/Debug.kt [hunk 6]
```diff
@@ -3,8 +3,10 @@
 import ftl.Main
 import ftl.run.exception.withGlobalExceptionHandling
 import picocli.CommandLine
+import java.io.File
 
 fun main() {
+    println(File("").absolutePath)
```
Comment (piotradamczyk5) [inline test_runner/src/test/kotlin/Debug.kt:9]: do we need this?

File: test_runner/src/test/kotlin/ftl/run/DumpShardsKtTest.kt [hunk 7]
```diff
@@ -183,24 +183,25 @@ class DumpShardsKtTest {
 
     @Test
     fun `dump shards obfuscated ios`() {
+        if (FtlConstants.isWindows) return // TODO Windows Linux subsytem does not contain all expected commands
```
Comment (piotradamczyk5) [inline test_runner/src/test/kotlin/ftl/run/DumpShardsKtTest.kt:186]:
```suggestion
        assumeFalse(FtlConstants.isWindows)
```

File: gradle.properties (hunk 1 above)
Comment (jan-gogo) [reply gradle.properties:16]: This flag makes command `./gradlew test` unusable, even after `./gradlew clean`.

File: test_runner/src/test/kotlin/Debug.kt (hunk 6 above)
Comment (jan-gogo) [reply test_runner/src/test/kotlin/Debug.kt:9]: Can be useful to determine if execution dir is a repository root or `test_runner` dir

File: test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt (hunk 4 above)
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt:37]: Pls, use GitHub's suggestion for convenience.

File: test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt (hunk 5 above)
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt:88]: I am not sure if this is a good idea. 
1. It cannot be simplified to `newFunction1() to newFunction2()` because `newFunction2()` will require result of `newFunction1()` as argument (shardChunks).
2. It will produce 2 additional functions. So we will have 3 tightly coupled private functions instead of just 1.
3.  This function:
```
private fun IosArgs.calculateConfigurationShards(
    targets: Map<String, List<String>>,
    regexList: List<Regex>
): Pair<List<Chunk>, List<XctestrunMethods>> = newFunction1() to newFunction2()
```
looks a little bit pointless.

File: test_runner/src/main/kotlin/ftl/ios/xctest/FindXcTestNamesV2.kt (hunk 2 above)
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/FindXcTestNamesV2.kt:25]: Yup why not.

File: test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV1.kt (hunk 3 above)
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV1.kt:16]: ok

File: test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt (hunk 4 above)
Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt:37]: it does not work well with multilines 😄

File: test_runner/src/test/kotlin/Debug.kt (hunk 6 above)
Comment (piotradamczyk5) [reply test_runner/src/test/kotlin/Debug.kt:9]: Maybe append "Running path: " at start of println message

File: test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt (hunk 4 above)
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/ReduceXcTestRunV2.kt:37]: > it does not work well with multilines 😄

Only for lines that are not changed in PR.

File: test_runner/src/test/kotlin/Debug.kt (hunk 6 above)
Comment (jan-gogo) [reply test_runner/src/test/kotlin/Debug.kt:9]: `Debug.kt` is only for debugging purpose

File: test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+data class XcTestRunData(
+    val rootDir: String,
+    val nsDict: NSDictionary,
+    val version: XcTestRunVersion = nsDict.getXcTestRunVersion(),
+    val shardTargets: Map<String, List<XctestrunMethods>> = emptyMap(),
+    val shardChunks: Map<String, List<Chunk>> = emptyMap()
+)
+
+fun IosArgs.calculateXcTestRunData(): XcTestRunData {
+    val xcTestRunFile = File(xctestrunFile)
```
Comment (pawelpasterz) [inline test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt:27]: I think it won't work for files provided via `gs://` path

**Review by pawelpasterz — CHANGES_REQUESTED** [review]
I think flank's support for files with GCS path is not working. Not sure if the bug was introduced within this PR though...
Please verify

File: test_runner/src/test/kotlin/Debug.kt (hunk 6 above)
Comment (piotradamczyk5) [reply test_runner/src/test/kotlin/Debug.kt:9]:
```suggestion
    println("Running path: ${File("").absolutePath}")
```

File: test_runner/src/main/kotlin/ftl/args/ValidateIosArgs.kt [hunk 9]
```diff
@@ -84,3 +86,32 @@ private fun IosArgs.validType() {
...
+                .shardTargets.values
+                .flatten()
+                .flatMap { it.values }
+                .flatten()
+
+            if (filteredMethods.isEmpty()) throw FlankConfigurationError(
+                "Empty shards. Cannot match any method to $testTargets"
+            )
+
+            if (filteredMethods.size < testTargets.size) {
```
[the remark this answers is not in the document]
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/args/ValidateIosArgs.kt:103]: If the filtered methods are less than the test targets specified in the config, we can notify the user about test targets that not match any method.

**Comment by axelzuziak-gogo** [comment]
Verified by running few sample projects with test plans. 

- [ ] Flank is correctly parsing xctestrun V2 file and splits it into separate ones using Test Configurations

- [ ] Flank correctly filters methods with Test-Targets flag

- [ ] Flank is missing the ability to specify test configuration, but it will be covered in separate ticket: #1379

File: test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt (hunk 8 above)
Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt:27]: That was not working previously - https://github.com/Flank/flank/blob/074cf0e1bc0afafd4dd99820e60df2fe427cf5a3/test_runner/src/main/kotlin/ftl/ios/Xctestrun.kt#L107

Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt:27]: This is the source of changes https://github.com/Flank/flank/commit/501ea710b7c949a294f4a35e75ccfa619655e90e

Comment (jan-gogo) [reply test_runner/src/main/kotlin/ftl/ios/xctest/XcTestData.kt:27]: We can implement support for `gs` scheme in another PR.

File: gradle.properties (hunk 1 above)
Comment (piotradamczyk5) [reply gradle.properties:16]: but we will disable cache for whole project.
I think that if there is some issue with `test` task then we should disable cache only for it

File: test_runner/src/main/kotlin/ftl/run/DumpShards.kt [hunk 10]
```diff
@@ -32,10 +33,22 @@ fun IosArgs.dumpShards(
...
+        it.value.flatMap { it.values }
+    }
+
+    val rawShards: Any = when (xcTestRunData.version) {
+        V1 -> xcTestRunShards.values.first()
+        V2 -> xcTestRunShards
+    }
+
+    val size = xcTestRunData.shardTargets.values
+        .flatten().size
```
Comment (axelzuziak-gogo) [inline test_runner/src/main/kotlin/ftl/run/DumpShards.kt:46]: Is this correct? I'm not sure if `size` will be correct for xctestrun V1. 
Should we return `rawShards.size` here?

**Review by pawelpasterz — APPROVED** [review]

**Review by piotradamczyk5 — CHANGES_REQUESTED** [review]
Please response to the cache and assume comment

**Review by adamfilipow92 — APPROVED** [review]

File: gradle.properties (hunk 1 above)
Comment (jan-gogo) [inline gradle.properties:16]:
```suggestion
```

**Review by piotradamczyk5 — APPROVED** [review]

**Review by axelzuziak-gogo — APPROVED** [review]
